### ISSUE TITLE
JS: Clicking "Debug" link above test results in "Error: Cannot find module '/home/[my_home]/node_modules/.bin/jest'"

### ISSUE THREAD

**Issue by haughki** [firsttris/vscode-jest-runner#266]
Type: <b>Bug</b>

WSL 2
Ubuntu 20.04.4 LTS
nvm 0.39.1
node v18.7.0
npm 8.15.0

Disabled all but these three VSCode extensions:
remote-wsl|ms-|0.66.3
vscode-remote-extensionpack|ms-|0.21.0
vscode-jest-runner|fir|0.4.48

To Reproduce
- Run ' npx create-react-app my-app'
- cd to 'my-app' directory
- Run 'code .'
- In VSCode, create my-app/src/supporting/MyClass.js based on this gist: https://gist.github.com/haughki/c7ab5d72a8357d211b1a5d6000f7b833#file-myclass-js
- In VSCode, create my-app/src/supporting/__test__/MyClass.test.js based on this gist: https://gist.github.com/haughki/0488d864794c539c7acfca757e01f280
- In VSCode, in MyClass.test.js, above the test declaration ( "test('testing', () => {" ) on l. 9, note the "Run|Debug" links
- Click "Debug"

Expected
The test runs in the terminal

Actual
A 'jest' terminal starts in the TERMINAL tab on the bottom pane. It tries to cd into the workspace dir before the command prompt is ready, and ends up in the home directory:
```
cd '/home/xxxxxx/projects/my-app'
Starting wsl-vpnkit service...
process already running.
xxxxxx:~$ 
```
Then, a "Debug Jest Tests" terminal starts, and runs the following commands which result in the error:

```
Starting wsl-vpnkit service...
process already running.
xxxxxx:~$  /usr/bin/env 'NODE_OPTIONS=--require /home/xxxxxx/.vscode-server/bin/6d9b74a70ca9c7733b29f0456fd8195364076dda/extensions/ms-vscode.js-debug/src/bootloader.bundle.js --inspect-publish-uid=http' 'VSCODE_INSPECTOR_OPTIONS={"inspectorIpc":"/tmp/node-cdp.29546-2.sock","deferredMode":false,"waitForDebugger":"","execPath":"/home/xxxxxx/.nvm/versions/node/v18.7.0/bin/node","onlyEntrypoint":false,"autoAttachMode":"always","mandatePortTracking":true,"fileCallback":"/tmp/node-debug-callback-2e4dfcdc9ab2401d"}' /home/xxxxxx/.nvm/versions/node/v18.7.0/bin/node node_modules/.bin/jest /home/xxxxxx/projects/my-app/src/supporting/__test__/MyClass.test.js -t testing --runInBand 
Debugger attached.
Waiting for the debugger to disconnect...
node:internal/modules/cjs/loader:959
  throw err;
  ^

Error: Cannot find module '/home/xxxxxx/node_modules/.bin/jest'
    at Module._resolveFilename (node:internal/modules/cjs/loader:956:15)
    at Module._load (node:internal/modules/cjs/loader:804:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v18.7.0
xxxxxx:~$ 
```

- I then install jest to my home dir:
xxxxxx:~$ npm install --save-dev jest
- Delete the existing 'jest' and 'Debug Jest Tests' from the VSCode TERMINAL tab on the bottom pane
- Set a breakpoint in MyClass.test.js on l. 10
- Above the test, click 'Debug' again

Expected
The test should run and break on the breakpoint

Actual
The terminal spawn again as above, and I get the following:
```
Starting wsl-vpnkit service...
process already running.
xxxxxx:~$  /usr/bin/env 'NODE_OPTIONS=--require /home/xxxxxx/.vscode-server/bin/6d9b74a70ca9c7733b29f0456fd8195364076dda/extensions/ms-vscode.js-debug/src/bootloader.bundle.js --inspect-publish-uid=http' 'VSCODE_INSPECTOR_OPTIONS={"inspectorIpc":"/tmp/node-cdp.2717-2.sock","deferredMode":false,"waitForDebugger":"","execPath":"/home/xxxxxx/.nvm/versions/node/v18.7.0/bin/node","onlyEntrypoint":false,"autoAttachMode":"always","mandatePortTracking":true,"fileCallback":"/tmp/node-debug-callback-01f3dde948ed1ced"}' /home/xxxxxx/.nvm/versions/node/v18.7.0/bin/node node_modules/.bin/jest /home/xxxxxx/projects/my-app/src/supporting/__test__/MyClass.test.js -t testing --runInBand 
Debugger attached.
```
The terminal just hangs at this point, the test never runs, breakpoint never hit.

VS Code version: Code 1.70.1 (6d9b74a70ca9c7733b29f0456fd8195364076dda, 2022-08-10T06:08:33.642Z)
OS version: Windows_NT x64 10.0.19042
Modes:
Remote OS version: Linux x64 5.4.72-microsoft-standard-WSL2
Remote OS version: Linux x64 5.4.72-microsoft-standard-WSL2

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8365U CPU @ 1.60GHz (8 x 1896)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.81GB (0.98GB free)|
|Process Argv|--file-uri=vscode-remote://wsl+Ubuntu/home/xxxxxx/projects/vsc_test_bug/NO_REG2.code-workspace --remote=wsl+Ubuntu --crash-reporter-id ea39c7bc-9acf-4ab0-a8de-b9cec0c6fcb0|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu|
|OS|Linux x64 5.4.72-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i5-8365U CPU @ 1.60GHz (4 x 1896)|
|Memory (System)|2.92GB (0.22GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu|
|OS|Linux x64 5.4.72-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i5-8365U CPU @ 1.60GHz (4 x 1896)|
|Memory (System)|2.92GB (0.22GB free)|
|VM|0%|
</details><details><summary>Extensions (3)</summary>

Extension|Author (truncated)|Version
---|---|---
remote-wsl|ms-|0.66.3
vscode-remote-extensionpack|ms-|0.21.0
vscode-jest-runner|fir|0.4.48

</details>

**Comment by firsttris** [comment]
maybe this happens because of your remote wsl connection?

can you use the vanilla debug feature from vscode. the addon is just executing that.

**Comment by haughki** [comment]
tl;dr
At this point, I don't _really_ see this as a Jest Runner issue. See details below. I _would_ like to see if I can sort out the weird terminal issue I describe below, and then post back to this thread. I can't remember if I can still post to the thread after you close the issue: if so, feel free to close it and I'll circle back when I get it sorted. Else, maybe you could leave open for a week -- I'll plan to figure this out in the next few days.


## Can you use the vanilla debug feature...
> can you use the vanilla debug feature from vscode. the addon is just executing that.

Yes. After adding a babel.config.js to the root:
```
module.exports = {
  presets: [
    ["@babel/preset-env", { targets: { node: "current" } }],
    "babel-preset-react-app",
  ],
}
```
I can attach the debugger and break running `:~/projects/my-app (master)$ ./node_modules/.bin/jest --runInBand /home/xxxxxx/projects/my-app/src/supporting/__test__/MyClass.test.js -t testing`
----------------------
## Well, I see now at least part of the problem...
The actual jest command:

`home/xxxxxx/.nvm/versions/node/v18.7.0/bin/node node_modules/.bin/jest /home/xxxxxx/projects/my-app/src/supporting/__test__/MyClass.test.js -t testing --runInBand`

is looking for `node_modules/.bin/jest` assuming the cwd is the project root, but the terminal is in my home dir. Hence: `Error: Cannot find module '/home/xxxxxx/node_modules/.bin/jest'` So, _I think that if the jest command was an abs path, it would fix the bug, sortof...._

But, Jest Runner _is_ already trying to `cd` the terminal into the project root; the `cd` just isn't happening 'in time'. This seems to have more to do with the 'start up' of the two terminals. Meaning:

- If there are no 'jest' or 'Debug Jest Tests' terminals already open
![image](https://user-images.githubusercontent.com/1500330/185233068-2492eede-1936-467a-9a02-761b113057be.png)
- when I click the 'Debug' link, it spawns first the 'jest' terminal, which ends up like:
![image](https://user-images.githubusercontent.com/1500330/185233110-bf59dd62-0219-4ccf-b54e-9d5927e6dc9c.png)
- And then a 'Debug Jest Tests' terminal that ends up like:
![image](https://user-images.githubusercontent.com/1500330/185234016-078d5904-35bf-4fff-9017-02553780630f.png)
- BUT if I then click on 'Debug' again, without first closing the two terminals, the debugger runs and hits the breakpoin. Note the `cd /home/xxxxxx/projects/my-app ; ` especially in the 'Debug Jest Tests' terminal:
![image](https://user-images.githubusercontent.com/1500330/185235266-ac3cfb8f-b334-4135-8321-dd7ad4a8f00b.png)
![image](https://user-images.githubusercontent.com/1500330/185235520-05117ee6-9d85-4d30-a815-fe76c0948a44.png)

So. It works, just not on the first run. I wonder why things are getting so munged on that first run: looks sorta like a timing thing -- that the terminals are taking too long to start. Maybe has something to do with my .bashrc that runs when the term opens (that's where the 'wsl-vpnkit' is coming from)...? But, if so, that seems like a bug (somewhere) in it's own right; i.e., a new terminal should be able to run a .bashrc to set itself up before whatever else happens, right?

**Comment by haughki** [comment]
## Solution
This was because of a line in my .bashrc:
```
# go home. This because of a bug in winterm settings: https://github.com/microsoft/WSL/issues/6995
~
```
I was telling my shell to cd to the home directory at the end of .bashrc because of an old WSL bug that I had 'fixed' with this tilde. Removing the line fixes the problem of the JS debugger looking for jest in the wrong place, etc.

Thank you for this fantastic plugin!!
